engine/script/src/script_vmath.cpp

```diff
@@ -2299,6 +2401,293 @@ namespace dmScript
         return 1;
     }
 
+    /*# creates a new quaternion from matrix4
+     *
+     * Creates a new quaternion with the components set
+     * according to the supplied parameter values.
+     *
+     * @name vmath.quat_matrix4
+     * @param matrix [type:matrix4] source matrix4
+     * @return q [type:quaternion] new quaternion
+     */
+    static int Quat_Matrix4(lua_State* L)
+    {
+        Matrix4* matrix = CheckMatrix4(L, 1);
+        PushQuat(L, dmVMath::Quat(matrix->getUpper3x3()));
+        return 1;
+    }
+
+    /*# creates a new matrix4 from translation, rotation and scale
+     *
+     * Creates a new matrix constructed from separate
+     * translation vector, roation quaternion and scale vector
+     *
+     * @name vmath.matrix4_compose
+     * @param translation [type:vector3|vecto4] translation
+     * @param rotation [type:quaternion] rotation
+     * @param scale [type:vector3] scale
+     * @return matrix [type:matrix4] new matrix4
+     * @examples
+     *
+     * ```lua
+     * local translation = vmath.vector3(103, -95, 14)
+     * local quat = vmath.quat(1, 2, 3, 4)
+     * local scale = vmath.vector3(1, 0.5, 0.5)
+     * local result = vmath.matrix4_compose(translation, quat, scale)
+     * print(result) --> vmath.matrix4(-25, -10, 11, 103, 28, -9.5, 2, -95, -10, 10, -4.5, 14, 0, 0, 0, 1)
+     * ```
+     */
+    static int Matrix4_Compose(lua_State* L)
+    {
+        Matrix4 translation_matrix = Matrix4::identity();
+        void* argument = 0;
+        const ScriptUserType translation_type = CheckUserData(L, 1, &argument);
+
+        if (translation_type == SCRIPT_TYPE_VECTOR3)
+        {
+            Vector3* translation = (Vector3*)argument;
+            translation_matrix.setTranslation(*translation);
+        }
+        else if (translation_type == SCRIPT_TYPE_VECTOR4)
+        {
+            Vector4* translation = (Vector4*)argument;
+            translation_matrix.setTranslation(translation->getXYZ());
+        }
+        else
+        {
+            return luaL_error(L, "%s.%s accepts (%s|%s) as first argument.", SCRIPT_LIB_NAME, "matrix4_compose",
+                SCRIPT_TYPE_NAME_VECTOR3, SCRIPT_TYPE_NAME_VECTOR4);
+        }
+        Matrix4 rotation_matrix = Matrix4::rotation(*CheckQuat(L, 2));
+        Matrix4 scale_matrix = Matrix4::scale(*CheckVector3(L, 3));
+        Matrix4 result = translation_matrix * rotation_matrix * scale_matrix;
+
+        PushMatrix4(L, result);
+        return 1;
+    }
+
+    /*# creates a new matrix4 from scale vector
+     *
+     * Creates a new matrix constructed from scale vector
+     *
+     * @name vmath.matrix4_scale
+     * @param scale [type:vector3] scale
+     * @return matrix [type:matrix4] new matrix4
+     * @examples
+     *
+     * ```lua
+     * local scale = vmath.vector3(1, 0.5, 0.5)
+     * local result = vmath.matrix4_scale(scale)
+     * print(result) --> vmath.matrix4(1, 0, 0, 0, 0, 0.5, 0, 0, 0, 0, 0.5, 0, 0, 0, 0, 1)
+     * ```
+     */
```

Comment:
This is a new thing for us I think, having explicit documentation entries for the same function.
I think it's borderline, as it might be tricky to keep the documentation in sync that way. The alternative may be visually awkward (for the arguments) thought.
However, the examples would be grouped together.

Good enough for now at least.

Reply:
Not new. Initially I was searching for example how to describe one function with different set of arguments and than found that https://github.com/defold/defold/blob/dev/engine/script/src/script_vmath.cpp#L1064-L1120. And did in the same way.